### ISSUE TITLE
Is incremental compilation actually supported?

### ISSUE THREAD

**Issue by neongreen** [tweag/rules_haskell#206]
I tried the following experiment:

* Take https://github.com/tweag/rules_haskell_examples/tree/master/vector
* `bazel build`
* Add a blank line to some module that has few reverse dependencies (e.g. `Data.Vector.Unboxed.Mutable`)
* `bazel build`

The time to build `vector` was the same, and, judging by the logs, all modules were recompiled. Am I doing something wrong?

My version of Bazel is 0.11.1.

**Comment by johnynek** [comment]
I think ghc is adding timestamps etc to the output, so it kills incrementality.

**Comment by johnynek** [comment]
@mboes I think there is something more here. If ghc is not reproducible, say each run puts a timestamp in the outputs or does not make deterministic outputs, then it won't matter how finely grained you make it, everything downstream will be rebuilt.

This is not an issue with other bazel supported languages (java, c++ native rules are reproducible, except in cases of macros, scala rules are reproducible).

It would be nice to document any known reproducibility issues with the current setup. Note, without reproducibility, caching is not going to work at all with these rules.

**Comment by johnynek** [comment]
https://ghc.haskell.org/trac/ghc/ticket/4012

There is the ticket about making ghc reproducible.

**Comment by mboes** [comment]
@johnynek agreed, GHC determinism would be nice. I gather @neongreen's issue is simpler than that: he defined one `haskell_library` and was expecting that trivial changes to only one module in the library would not cause the whole library to be recompiled. We don't track module-level dependencies in Bazel, so that's why everything gets recompiled. Now, whether downstream dependencies (not present in the original repro of this ticket) get recompiled because the output of `haskell_library` spuriously changes when a trivial source-level change is made, that's another issue. But one that's entirely dependent on upstream AFAICT?

**Comment by johnynek** [comment]
@mboes yes, I think you are right.

I didn’t look in detail at how the rules are written, but when I tried to take a stab at writing Haskell rules before I tried to see if I could only pass the interface files as inputs to haskell_library and then in binary you would do the linking. If this is how it worked, you could imagine it is easier to make the interface files deterministic, or even post process them to make them deterministic. So, it might not entirely be an upstream issue.

Note, in Java google make ijar, a tool to remove the code elements from a jar so you only keep the non-private type signatures. There may be an analogous approach we could do here which would all minor changes to not cause total downstream recompilation.
